File: orchestrator-java-reactive/src/main/java/com/github/mzr/coder/orchestrator/java/reactive/service/impl/DecisionServiceImpl.java
package com.github.mzr.coder.orchestrator.java.reactive.service.impl;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.mzr.coder.model.ServiceResponse;
import com.github.mzr.coder.orchestrator.java.reactive.config.UrlConfig;
import com.github.mzr.coder.orchestrator.java.reactive.service.DecisionService;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class DecisionServiceImpl implements DecisionService {

    private final WebClient.Builder webClientBuilder;

    private final UrlConfig urlConfig;

    private final ObjectMapper objectMapper;

    @SneakyThrows
    @Override
    public Mono<List<String>> getStrategy(String name) {
        log.info("Trying get strategy for request with name {}", name);
        return webClientBuilder.baseUrl(urlConfig.getDecisionService())
                .build()
                .get()
                .uri("/strategy")
                .retrieve()
                .bodyToMono(String[].class)
                .map(array -> Arrays.stream(array).collect(Collectors.toList()))
                .doOnSuccess(services -> {
                            try {
                                log.info("Got strategy for request with name {}. List services {}",
                                        name,
                                        objectMapper.writeValueAsString(services)
                                );
                            } catch (JsonProcessingException e) {
                                throw new RuntimeException(e);
                            }
                        }
                );
    }

    @Override
    public Mono<ServiceResponse> getDecision(String name, List<ServiceResponse> serviceResponses) {
        var monoBody = Mono.just(serviceResponses)
                .doOnSuccess(x -> log.info("Trying get decision for request with name {}", name));
        return webClientBuilder.baseUrl(urlConfig.getDecisionService())
                .build()
                .post()
                .uri("/decision")
                .body(monoBody, List.class)
                .retrieve()
                .bodyToMono(ServiceResponse.class)
                .doOnSuccess(sr -> log.info("Got decision for request with name {}, approve {}", name, sr.getApprove()));

    }
}

File: orchestrator-kotlin-coroutine/src/main/java/com/github/mzr/coder/orchestrator/kotlin/reactive/service/impl/DecisionServiceImpl.kt
package com.github.mzr.coder.orchestrator.kotlin.reactive.service.impl

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.mzr.coder.model.ServiceResponse
import com.github.mzr.coder.orchestrator.kotlin.reactive.config.UrlConfig
import com.github.mzr.coder.orchestrator.kotlin.reactive.service.DecisionService
import kotlinx.coroutines.reactor.awaitSingle
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import java.util.*

@Service
class DecisionServiceImpl(
    private val webClientBuilder: WebClient.Builder,
    private val urlConfig: UrlConfig,
    private val objectMapper: ObjectMapper
) : DecisionService {

    private val log = LoggerFactory.getLogger(DecisionServiceImpl::class.java)

    override suspend fun getStrategy(name: String): List<String> {
        log.info("Trying get strategy for request with name {}", name)
        val services = webClientBuilder.baseUrl(urlConfig.decisionService)
            .build()
            .get()
            .uri("/strategy")
            .retrieve()
            .bodyToMono(Array<String>::class.java)
            .awaitSingle()
        log.info(
            "Got strategy for request with name {}. List services {}",
            name,
            objectMapper.writeValueAsString(services)
        )
        return services.toList()
    }

    override suspend fun getDecision(name: String, servicesResponse: List<ServiceResponse>): ServiceResponse {
        log.info(
            "Trying get decision for request with name {}",
            name
        )
        val serviceResponse = webClientBuilder.baseUrl(urlConfig.decisionService)
            .build()
            .post()
            .uri("/decision")
            .bodyValue(servicesResponse)
            .retrieve()
            .bodyToMono(ServiceResponse::class.java)
            .awaitSingle()
        log.info("Got decision for request with name {}, approve {}", name, serviceResponse.approve)
        return serviceResponse
    }
}